lib/src/clojure_lsp/dep_graph.clj

```diff
@@ -0,0 +1,344 @@
+(ns clojure-lsp.dep-graph
+  (:require
+   [clojure-lsp.shared :as shared]
+   [clojure.set :as set]))
+
+;;;; Syntax
+
+;; () grouping
+;; | choice
+;; ? zero or one
+;; + one or more
+;; * zero or more
+;; '' literal
+;; {} map {k1 v1 ...}
+;; #{} set #{i1 ...}
+;; #<> multiset #<i1 ...>
+;;     a mulitset is represented internally as a hashmap {i1 f1 ...} where f1 =
+;;     integer > 0, frequency of i1
+
+;; ns = symbol
+;; filename = string
+;; alias = symbol
+;; lang = ':clj' | ':cljs' | ':edn'
+
+;; dep-graph-item =
+;; {(':dependencies' #<ns*>)?
+;;  (':dependents' #<ns*>)?
+;;  (':aliases' #<(alias | nil)*>)? ;; nil for when the ns is required without an alias
+;;  (':filenames' #{filename*})?
+;;  (':internal?' boolean)?
+;;  (':from-internal?' boolean)?
+;;  (':from-langs' #<lang*>)?}
+
+;; :dependencies is a multiset of the namespaces that this namespace depends on.
+;; It's a multiset because a dependency may be required several times, either
+;; with different aliases, or because the dependent namespace is defined across
+;; several files. We keep a multiset because removing one of these duplicates
+;; doesn't erase the dependency/dependent relationship. If a namespace has no
+;; dependencies, the :dependencies will either be absent from the
+;; dep-graph-item, or be an empty multiset. The system doesn't make guarantees
+;; about which it will be.
+
+;; :dependents is a multiset of the namespaces that depend on this namespace.
+;; May be either empty or absent if namespace isn't depended on by other
+;; namespaces.
+
+;; If the code had `(ns aaa (require [bbb :as b]))`, then `'bbb` would be in
+;; `'aaa`'s :dependencies, and `'aaa` would be in `'bbb`'s :dependents.
+
+;; :aliases is a multiset of the symbols that are used to refer to this
+;; namespace. If the code had `(:require [aaa :as a])`, then `'aaa` would
+;; include `'a` in its aliases. Unaliased requires like `(:require [aaa])` will
+;; include `nil` in their aliases. May be either empty or absent if a namespace
+;; isn't required by other namespaces.
+
+;; :filenames is a set of the names of the files in which the namespace is
+;; defined. This can be either absent or empty if the namespace is required but
+;; never defined, either as a syntax error or via :as-alias.
+
+;; :internal? is true when the namespace is defined in any internal file, falsy
+;; otherwise. May be absent if the namespace is required but never defined, as
+;; for :filenames.
+
+;; :from-internal? is true when the namespace is required by any internal
+;; namespace, falsy otherwise. May be absent if the namespace isn't required by
+;; other namespaces.
+
+;; :from-langs is a multiset of the langs of the :namespace-usages or files (if the
+;; :namespace-usages don't have langs) that use this namespace. May be either
+;; empty or absent if the namespace isn't required by other namespaces.
+
+;; file-meta-item =
+;; {(:namespaces #{ns*})?
+;;  :internal? boolean
+;;  :langs #{lang+}}
```

Comment:
I think `file-metas` is similar to a already existing `:documents` in the db? would make sense to use `:documents` but add those extra infos there?

Reply:
I hadn't thought of that, but maybe... let me think about it more. One of the differences between `:file-meta` and `:documents` is that `:documents` is keyed by URI. But I'm considering your suggestion about re-keying by URI more seriously. I'll detail those thoughts in a separate comment.

As an aside... I'm really excited about your idea of caching the rewrite-clj parse. We all spend much more time reading code than writing it, so the parses aren't changing most of the time. I think having the parse cached would speed up things like `:hover`, which need a parse but which are invoked as the user navigates, not as they type. Now that I'm thinking about `:documents`, that's probably a good place to cache the parse. You could experiment with putting it there independently of when I merge dep-graph.

Reply:
See other comment for details, but I moved the file metadata into :documents, as suggested. Much nicer. 💯